store/tikv/2pc.go

```diff
@@ -386,8 +419,29 @@ func (c *twoPhaseCommitter) prewriteSingleBatch(bo *Backoffer, batch batchKeys)
 				return errors.Trace(err1)
 			}
 			log.Debugf("con:%d 2PC prewrite encounters lock: %v", c.connID, lock)
-			locks = append(locks, lock)
+			if lock != nil {
+				locks = append(locks, lock)
+			}
+
+			// If there is any contract error, it must be bugs in TiDB!
+			// TODO: Add prometheus metrics and alert here?
+			if notExist := keyErr.GetNotExist(); notExist != nil {
+				ex := c.mutations[string(notExist.Key)]
+				if ex.hasContract {
+					log.Error(errors.ErrorStack(errors.Trace(errors.Errorf("CONTRACT BUG!!! The key %v should have value, but not found\n", notExist.Key))))
+				}
+			}
+			if exist := keyErr.GetAlreadyExist(); exist != nil {
+				ex := c.mutations[string(exist.Key)]
+				if ex.hasContract {
+					log.Error(errors.ErrorStack(errors.Trace(errors.Errorf("CONTRACT BUG!!! The key %v must not have value, but value found\n", exist.Key))))
```

Comment:
If we skip the check lazy condition pair step, it would be normal to have this error and we should return this error to the client.